src/Functions/IFunction.h

```diff
@@ -208,6 +208,29 @@ class IFunctionBase
       */
     virtual bool hasInformationAboutMonotonicity() const { return false; }
 
+    /** Function is called "short-circuit" if it's arguments can be evaluated lazily
+      * (examples: and, or, if, multiIf). If function is short circuit, it must
+      *  implement method executeShortCircuitArguments for lazy arguments execution,
+      *  this method will be called before function execution.
+      */
+    virtual bool isShortCircuit() const { return false; }
+
+    /** Should we evaluate this function lazily in short-circuit function arguments?
+      * If function can throw an exception or it's computationally heavy, then
+      * it's suitable, otherwise it's not (due to the overhead of lazy execution).
+      * Suitability may depend on function arguments.
```

Comment:
I think, we must evaluate all arguments as short-circuit in case user asks it. Otherwise it would be very unusual.
Also I don't think we can detect somehow if short-circuit variant is faster then usual one.

Probably, we don't need this?

Reply:
We discussed it with @alexey-milovidov. For simple functions filtering column costs almost the same as execution the function on the full column. So, performance of such function with short-circuit execution could be much worse then without it. We decided to mark all functions as suitable for short-circuit execution or not. I tried to mark